### ISSUE TITLE
Hiding Notes to the macOS dock doesn't seem to change its visibility state

### ISSUE THREAD

**Issue by guihkx** [nuttyartist/notes#458]
After "closing" Notes to the macOS dock, somehow Qt still thinks the window is visible (verifiable by calling [`isVisible()`](https://doc.qt.io/qt-5/qwidget.html#visible-prop)).

The side effect of this is having to double-click the system tray icon to restore Notes, instead of just clicking it once, which is what the intended behavior should be.

---

**Steps to reproduce:**

1. Close Notes to dock (by clicking on the red close button)
2. Click on Notes' tray icon only once

**What should happen:**

Notes' window should be restored.

**What happens instead:**

Nothing, because Qt still thinks the window is visible at this point, so this single click ends up hiding the window again, therefore requiring yet another click to actually restore the window.

**Bug in action:**

https://user-images.githubusercontent.com/626206/214741606-ec0ce560-194e-4250-8188-c4dbe2ff0cdc.mp4

**Comment by bjorn** [comment]
We may want to use [`isActiveWindow`](https://doc.qt.io/qt-5/qwidget.html#isActiveWindow-prop) instead to decide whether we want to show or hide the window, when the tray icon is clicked. This could have the added benefit, that the window will be brought to the front, when it is visible but obscured by other windows.

**Comment by guihkx** [comment]
Oh, nice. I'll experiment with that.

>This could have the added benefit, that the window will be brought to the front, when it is visible but obscured by other windows.

I'm not sure this will work, most OSes prevent us from stealing focus from other windows. There are workarounds, however...

**Comment by guihkx** [comment]
Okay, even [`isActiveWindow()`](https://doc.qt.io/qt-5/qwidget.html#isActiveWindow-prop) returns `true` even though the window is clearly not active:

https://user-images.githubusercontent.com/626206/216593922-e1dd1ec9-f47e-4ed2-b068-622e749b4ca9.mp4

The code I used for a quick test:

```diff
diff --git a/src/mainwindow.cpp b/src/mainwindow.cpp
index 6b1ab86..1b3d011 100644
--- a/src/mainwindow.cpp
+++ b/src/mainwindow.cpp
@@ -190,6 +190,7 @@ void MainWindow::InitData()
 void MainWindow::onSystemTrayIconActivated(QSystemTrayIcon::ActivationReason reason)
 {
     if (reason == QSystemTrayIcon::Trigger) {
+        qDebug() << "isActiveWindow():" << isActiveWindow();
         setMainWindowVisibility(!isVisible());
     }
 }
```

If I had to guess, we'll have to probably emit some signal to Qt from the Objective-C code after we hide the window here:

https://github.com/nuttyartist/notes/blob/fb7e449355f770c9affeab2994a0ed3cf3261d0a/src/framelesswindow.mm#L33-L37

I have no clue how we'd do that, though.

**Comment by bjorn** [comment]
> If I had to guess, we'll have to probably emit some signal to Qt from the Objective-C code after we hide the window here:

That looks likely, since we're apparently hiding the window without involving Qt. Why would we not just call Qt's own `hide()` function instead? Judging by the implementation of `zoomButtonAction` we could just do:

```objc
    Q_UNUSED(sender);
    if (0 == self.window) return;

    self.window->hide();
```

**Comment by guihkx** [comment]
I tried a variation of that (I had to replace `+` by `-` in the function definition, otherwise it wouldn't compile):

```patch
diff --git a/src/framelesswindow.mm b/src/framelesswindow.mm
index e0f2c54..7aeb302 100644
--- a/src/framelesswindow.mm
+++ b/src/framelesswindow.mm
@@ -24,16 +24,18 @@
 @interface ButtonPasser : NSObject{
 }
 @property(readwrite) CFramelessWindow* window;
-+ (void)closeButtonAction:(id)sender;
+- (void)closeButtonAction:(id)sender;
 - (void)zoomButtonAction:(id)sender;
 @end
 
 @implementation ButtonPasser{   
 }
-+ (void)closeButtonAction:(id)sender
+- (void)closeButtonAction:(id)sender
 {
     Q_UNUSED(sender);
-    [NSApp hide:nil];
+    if (0 == self.window) return;
+
+    self.window->hide();
 }
 - (void)zoomButtonAction:(id)sender
 {    
```

... But it didn't work:

https://user-images.githubusercontent.com/626206/217714952-097b4108-b5f8-4bc5-ac22-f7b3d5534855.mp4

The error upon clicking the 'close' button:

```
Notes[1855:32142] +[ButtonPasser closeButtonAction:]: unrecognized selector sent to class 0x1046bf7d0
```

Perhaps that's why the [original implementation](https://github.com/Bringer-of-Light/Qt-Nice-Frameless-Window/blob/3002bf536e09a8fe01bbb03e441dde2d7ff4883a/framelesswindow/framelesswindow.mm#L33-L39) used     `GetFrontProcess()` / `ShowHideProcess()` too... 😕 

Nevertheless, I don't see why this wouldn't be possible, it just needs a little more research on the topic.

**Comment by bjorn** [comment]
@guihkx Maybe we need this code regarding the `passer` also for the close button action, to make sure `self.window` is set:

https://github.com/nuttyartist/notes/blob/1142bce844839c6757b2c94f3c7ffa6a92494844/src/framelesswindow.mm#L86-L91

Currently the code for the close button action looks like:

https://github.com/nuttyartist/notes/blob/1142bce844839c6757b2c94f3c7ffa6a92494844/src/framelesswindow.mm#L115-L116

**Comment by guihkx** [comment]
Dear god, I went into an ugly Objective-C rabbit hole. :)

The good part is that I learned some things while doing this, and I was able to fix this bug in particular.

The bad part is that it revealed some other bugs and even introduced a new one.

I'll try to iron things out before sending a pull request...

**Comment by guihkx** [comment]
I'm kind of stuck now, unfortunately...

This bug in particular I was able to fix, and here's my approach: https://github.com/guihkx/notes/commit/0da5ab0ba0d7ef252821f04e1233b5811fad454e

The idea behind it is to hook into AppKit's [`NSWorkspaceDidHideApplicationNotification`](https://developer.apple.com/documentation/appkit/nsworkspacedidhideapplicationnotification) and [`NSWorkspaceDidUnhideApplicationNotification`](https://developer.apple.com/documentation/appkit/nsworkspacedidunhideapplicationnotification) events, and when they're fired, I just manually call Qt's `hide()` / `show()` from the Objective-C code.

I'm not sure if that's the best (or even the correct) approach, but on my tests it worked quite consistently...

---

However, I noticed a different (yet similar) bug:

Clicking on the tray icon will toggle main window visibility, but since that's also done with Qt's `hide()` / `show()`, it ends up causing similar problems. For example, not being able to restore Notes simply by clicking on its icon on the dock (after you hid it by clicking on the tray icon)... 😠

So then I had a big 🧠 moment: What if I try a similar approach, and hook into Qt's `hideEvent()` / `showEvent()`? And from there, I could manually call AppKit's `hide:` / `unhide:`, right?

And that's what I tried here: https://github.com/guihkx/notes/commit/2bdc3ca189a216b7f1f7feef433dcc8244f1cf86

However, that only works for the first icon toggle (for whatever reason???), and I have no idea what to try now...

**Comment by guihkx** [comment]
>However, that only works for the first icon toggle (for whatever reason???)

Uhhh... So I grabbed both [build artifacts](https://github.com/guihkx/notes/actions/runs/4177067661) and ran those on my Big Sur VM, and it works all of a sudden?????

It still doesn't work properly on my local build, though. So I don't know what the hell is going on here.

@nuttyartist Can you please test the artifact build I linked here? Just run the build using the terminal, and then keep clicking on the tray icon; If it's working correctly, a message should *always* pop up on the terminal...

**Comment by nuttyartist** [comment]
Well, it does not always show up. Not very consistent.
![test](https://user-images.githubusercontent.com/16375940/219039733-5307c8f8-79e9-431a-b6d7-2db876b24fab.gif)

**Comment by guihkx** [comment]
Oh, wow. Yours appears to be less consistent, for some reason... :(

On my Big Sur VM it's "perfect" (using the build generated by GitHub, anyway):

https://user-images.githubusercontent.com/626206/219141307-d1850635-e9d2-43d2-8b65-d77af561a32d.mp4

But yeah, the first commit works perfectly here. The second one doesn't seem to be consistent for reasons beyond me, but in my opinion it's still better than what we currently have...

**Comment by nuttyartist** [comment]
> but in my opinion it's still better than what we currently have...

Fair. Let's go with that for now.

**Comment by guihkx** [comment]
The changes of #478 were manually merged into `dev`. I somehow messed up that PR badly.
